Add support for select, select!, reject, reject! and remove delete_if for Hash, and Array

devdazed opened the pull request:
This PR adds support for `select`, `select!`, `reject`, and `reject!` to `Hash, and extends the functionality of `Array`.  It also removes `delete_if` as it is too close of as alias to `reject!`  

This is primarily getting my feet wet with contributing to the project.  I would like to ensure that I am following all the proper guidelines and practices so that I can contribute to larger parts of the project in the future.

Edit: for support of the above methods, `each_with_object` is added.

jhass commented on src/hash.cr:
```diff
@@ -272,6 +272,26 @@ class Hash(K, V)
+  end
+
+  def select!(&block : K, V -> U)
+    delete_if{|k, v| !yield k, v }
```
Minor style note, I personally strongly prefer a space before the `{` ;)

devdazed replied:
ok, is there a style guide somewhere?  I want to be sure I am coding within the proper guidelines.

jhass replied:
We don't have one (yet), thus I carefully used "personally" ;D

ysbaddaden replied:
Maybe we should have one, based on https://github.com/bbatsov/ruby-style-guide

jhass replied:
I just link https://gist.github.com/jhass/a5ae80d87f18e53e7b56 instead of saying what I think of that one ... :P

Nakilon replied:
1. *space before `{`* where? in `delete_if{`? I don't like space there
2. Never ever base any style-guide on bbatsov thing, because that project was mental from the beginning -- full of random nonsense, that influenced others a lot, because more clever guys were not ready to tell others how to write code, but these illogical amateurs were.

dbackeus replied:
The "I like" and "I don't like" is essentially the problem that a style guide solves.

For now, until there is an official guide, for this case it is simple enough to look at the general coding style used in the project (search code for `map {` vs `map{` etc). Which is clearly in favour of `delete_if { |k, v| !yield k, v }`, that is space both before `{` and `|`.

Regarding taking things from bbatsov as a base isn't necessarily a bad idea. I've yet to see any ruby style guide that isn't more or less a subset of what's in there. A lot of what it does is mention the different ways things can be done and then argue that you should be choosing one consistent style for your own project.

Nakilon replied:
`I like` and `I don't like` isn't a problem, but the opposite. Programming isn't a religion, that should have a Bible, that everyone should follow. Rules shouldn't be created as quickly as possible just to exist and make programmer harder to write own code. Take a look at C/C++ -- they still have several styles at the same time because it's obvious that there shouldn't be the single one. Crystal as any another language needs years and the style-guide should be written but the most experienced guys, not like bbatsov or his backyard friends.

jhass replied:
Gosh, if I had known what offspring this would take. Please let's not pollute this PR further with completely OT discussion about styleguides, mkay?

jhass commented on src/hash.cr:
```diff
@@ -272,6 +272,26 @@ class Hash(K, V)
 
+  def select(&block : K, V -> U)
+    inject({} of K => V) do |memo, k, v|
+      yield(k, v) ? memo.merge!({k => v}) : memo
```
`each_with_object` should allow to get rid of the ternary in favor of a modifier-if.

devdazed replied:
`each_with_object` doesn't seem to exist on `Hash`

jhass replied:
Oh. Mmh, would be an easy enough addition, no? ;)

devdazed replied:
sure, `Ruby` has this defined on `Enumerable` does it make sense to put it there and have `Hash` inherit from it?

jhass replied:
Our Enumerable has it already in fact. Atm we can have Hash include Enumerable because there's no autosplatting and thus its `#each` becomes incompatible with what Enumerable demands.

bcardiff replied:
More in https://github.com/manastech/crystal/issues/798#issuecomment-112812706

jhass commented:
Looking good, got two minor style remarks.

Would you mind adding docs? I know that the current class is largely undocumented, but it'd be nice if at least all new stuff adds documentation, at least until we have less backlog on documenting existing stuff ;)

devdazed commented:
I'll add docs. I've seen most docs in a Markdown format.  Is there a specific doc style to follow?  Like rdoc or something, or is it just Markdown?

jhass commented:
It's just markdown

PragTob commented:
Do we really want to add individual implementations for these methods when there is a plan to change parsing so we can decompose {k,v} tuples so that they can be used by Enumerable and hence just include `Enumerable` in Hash?

(see #105 and [asterites comment for more background](https://github.com/manastech/crystal/issues/105#issuecomment-131547103))

jhass commented:
Yes, we can always drop them when that actually happens. Besides we'll need the self modifying variants on Hash anyway.

asterite commented:
Hash#select returns a Hash, Enumerable#select returns an Array. This is not a shared implementation from Enumerable.

PragTob commented:
doh, right sorry

asterite commented:
Is there something missing before merging this? We can add docs later if that's what preventing the merge.

jhass commented:
I'd really love to see `each_with_object` added and then these implemented with it, using `inject` here is just misuse and we shouldn't add bad examples to stdlib IMO.

jhass commented:
@devdazed ping, still interested in getting this in, any time yet?

devdazed commented:
Yeah, sorry, been really busy at work. I'll try to hammer this out in the next couple of days.






On Mon, Sep 21, 2015 at 2:40 PM -0700, "Jonne Haß" <notifications@github.com> wrote:












@devdazed ping, still interested in getting this in, any time yet?



—
Reply to this email directly or view it on GitHub.

jhass commented:
Sounds great!

devdazed commented:
ok, let me know if anything else needs to change.

jhass commented on src/hash.cr:
```diff
@@ -278,6 +290,35 @@ class Hash(K, V)
+    end
+  end
+
+  # Equivalent to `Hash#delete_if`, but returns nil if no changes were made.
```
The comment is wrong. Also while we try to avoid aliases, I personally prefer `reject!` over `delete_if` too. What do others think?

devdazed replied:
hmm, are you saying change the code to not use `delete_if` or just change the docs to not reference it?

jhass replied:
It does not return `nil` if no changes were made (which the current doc claims). Thus it's identical to `delete_if`, thus it's an alias.

devdazed replied:
hmm, my intention was for it to return nil. so it acts in the same way as Ruby's `reject!`. let me look at it

devdazed replied:
ok, fixed the methods to act the way they should and not be direct aliases. also added tests for it.  I needed the `length` of the hash to do it so I also added that method as well.

devdazed replied:
ok, fixed the methods to act the way they should and not be direct aliases. also added tests for it. anything else you can see here?

jhass replied:
So what do others think? Do we want `reject!` or is it too similar to `delete_if` and one should be removed/left out?

asterite replied:
We can probably just keep `reject!`, and remove `delete_if` (also from Array), no? `select/reject` are known opposites in Ruby.

devdazed replied:
I personally have never used `delete_if` in Ruby.  I would be just fine removing it. Although the use-cases are slightly different, I agree they are very similar.

jhass replied:
Yeah, let's drop `delete_if`. Also from `Array` after ensuring `Array#reject!` is there.

devdazed replied:
it looks like `Array` has `select` and `reject` however it doesn't quack the same way Ruby's does where `nil` is returned when no changed have been made.  This is a very nice feature that allows for things like 

```rb
if arr.reject!{ |e| e.foo == :bar }
  # do something if there was a change
end
```

Should I make them all work the same way?

jhass replied:
I never used that in Ruby and never had the urge to, but I guess it makes a false friend otherwise, so I'd say yes.

asterite replied:
I never used that either, and right now all bang methods return self, so we are already not quacking like Ruby in many places. I wouldn't mind returning self here.

asterite replied:
@devdazed Did you use that feature in real code?

devdazed replied:
this specific syntax, no, but i have been known to do something like
```rb
unchanged = foo.reject!( ... )
do_something unless unchanged
```

which can also be written like
```rb
new_foo = foo.reject( ... )
unchanged = new_foo.size == foo_size
do_something unless unchanged
```

which is arguably cleaner, albeit less performant.

devdazed replied:
my question here is, how compatible with Ruby does Crystal want to be?

asterite replied:
I guess we could return `nil`. My question was more: once you know it changed or it didn't change, what "common" actions are done? What's a real example for `do_something`? Also, getting the previous length and comparing it to the new length might work and should have the same performance.

The reason I like bang methods to return self is that you can chain calls, but if the `nil` value is something pretty common to use, we can keep that functionality.

jhass replied:
Avoiding false friends when opportune is nice, but compatibility is no requirement in forming a decision, just an argument.

will replied:
If these do return nil, then it makes chaining a lot harder, since it makes the return type a union, right?

devdazed replied:
well, i would typically shy away from chaining when mutating the object and instead chain on the `reject`, like `foo.select{ .. }.reject{ .. }`, personal preference maybe. I like to keep from using the mutation methods except in exceptional cases.

will replied:
Ok, yeah that makes sense.

jhass commented on src/hash.cr:
```diff
@@ -211,12 +211,29 @@ class Hash(K, V)
+  # Returns the number of KV pairs in the hash
+  def length
+    inject(0) { |m, _k, _v| m += 1 }
+  end
```
We dropped all `length` methods in favour of their `size` alias.

devdazed replied:
ah, will fix

devdazed commented:
ok these 3 commit do the following
 - remove `Hash#delete_if`
 - change `Array#reject!` and `Array#select!` to act the way Ruby's does. Added `Array#select` and `Array#reject`. Deprecated `Array#delete_if`
 - Removed `Array#delete_if`, Changed compiler references to point to `reject!`

One question, `delete_if` is still defined in `SimpleHash`. Should we give `SimpleHash` the `select` and `reject` methods and remove `delete_if` there? It is also defined in `src/compiler/crystal/semantic/dependencies.cr`, do we want to rename that method as well, to be `reject!` ?

jhass commented:
Doesn't Array get `select` and `reject` from `Enumerable`?

devdazed commented:
You're right. my apologies. Still new to the codebase.

devdazed commented:
ok removed them.

jhass commented:
I'd say go all the way, slay `delete_if` everywhere and replace it with `reject!` (and ideally `select!`) with the same semantics everywhere. Consistency wins over preference :)

devdazed commented:
ok, let me know if you would like to to reduce it more or if this is good.

jhass commented:
LGTM :+1:

jhass commented:
Thank you!
